enclave/src/bpf/enclave.bpf.c

```diff
@@ -0,0 +1,299 @@
+#include "vmlinux.h"
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <error.h>
+
+#define EPERM 1
+#define TASK_COMM_LEN 16
+
+/*
+ * Maps and related structures
+ * ===========================
+ *
+ * All structs below are either BPF maps or structures used as values of those
+ * maps.
+ */
+
+/*
+ * runtimes - BPF map containing the process names of container runtime init
+ * processes (for example: `runc:[2:INIT]` which is the name of every init
+ * process for runc).
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 16);
+	__type(key, u32);
+	__type(value, u32);
+} runtimes SEC(".maps");
+
+enum container_policy_level {
+	POLICY_LEVEL_LOOKUP_ERR = -2,
+	POLICY_LEVEL_NOT_FOUND = -1,
+
+	POLICY_LEVEL_RESTRICTED,
+	POLICY_LEVEL_BASELINE,
+	POLICY_LEVEL_PRIVILEGED
+};
+
+struct container {
+	enum container_policy_level policy_level;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct container);
+} containers SEC(".maps");
+
+struct process {
+	u32 container_id;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u64);
+	__type(value, struct process);
+} processes SEC(".maps");
+
+/*
+ * Utils
+ * =====
+ *
+ * Util functions.
+ */
+
+/*
+ * hash - simple string hash function which allows to use strings as keys for
+ * BPF maps even though they use u32 as a key type.
+ * @str: string to hash
+ * @len: length of the string
+ *
+ * Return: an u32 value representing the given string.
+ */
+static __always_inline u32 hash(const char *str, size_t len)
+{
+	u32 hash = 0;
+	int i;
+
+	for (i = 0; i < len; i++)
+		hash += str[i];
+
+	return hash;
+}
+
+/*
+ * task_pid - find a PID of the given process
+ * @task: task to find a PID for
+ *
+ * Return: a PID.
+ */
+static __always_inline u64 task_pid(struct task_struct *task)
+{
+	return (u64) BPF_CORE_READ(task, tgid) << 32 | BPF_CORE_READ(task, pid);
+}
+
+/*
+ * add_container - register the container and its init process in BPF maps
+ * @task: the init process of the container runtime
+ *
+ * Return: 0 if all BPF map operations are successful, otherwise an error code.
+ */
+static __always_inline int add_container(struct task_struct *task)
+{
+	int ret;
+
+	u32 container_id = bpf_get_prandom_u32();
+	struct container container = {
+		.policy_level = POLICY_LEVEL_BASELINE
+	};
+
+	ret = bpf_map_update_elem(&containers, &container_id, &container, 0);
+	if (ret < 0)
+		return ret;
+
+	u64 pid = task_pid(task);
+	struct process process = {
+		.container_id = container_id
+	};
+
+	ret = bpf_map_update_elem(&processes, &pid, &process, 0);
```

Comment:
Could this also return `-1` and set `errno`?

Reply:
It's a bit confusing, but:

* `bpf_map_update_elem` in **userspace** sets `errno`
* `bpf_map_update_elem` in **BPF program** doesn't set errno

Here we are in the BPF context, so `errno` is not used.